### PR TITLE
Add relations export option to package alg

### PR DESCRIPTION
**Pull request by elpaso**
Adds an option to the package algorithm to follow project relations and export related tables following the hierarchy upstream and downstream, taking selected features into account.

immagine

Funded by: ARPA Piemonte

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmpackage.cpp [hunk 1]
```diff
@@ -74,14 +75,197 @@ QgsPackageAlgorithm *QgsPackageAlgorithm::createInstance() const
...
+              while ( it.nextFeature( selectedFeature ) )
+              {
+                QgsFeature referencedFeature { relation.getReferencedFeature( selectedFeature ) };
+                if ( referencedFeature.isValid() )
+                {
+                  selected.insert( referencedFeature.id() );
+                }
+              }
+              relatedLayer->selectByIds( selected, Qgis::SelectBehavior::AddToSelection );
+            }
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmpackage.cpp:159]: This is expensive code to run in the main thread (blocking). Can you move it to the processAlgorithm call instead?

File: src/analysis/processing/qgsalgorithmpackage.cpp [hunk 2]
```diff
@@ -74,14 +75,197 @@ QgsPackageAlgorithm *QgsPackageAlgorithm::createInstance() const
...
+          // in that case the user explicitly marked the layer for export and it is supposed to be exported fully.
+          QgsFeatureIds selected;
+
+          if ( onlySaveSelected && ( ! layerWasExplicitlyAdded || referencingLayer->selectedFeatureCount() > 0 ) )
+          {
+            if ( ! layers.contains( qobject_cast<QgsMapLayer *>( referencingLayer ) ) || referencingLayer->selectedFeatureCount() > 0 )
+            {
+              QgsFeatureIterator it { vLayer->getSelectedFeatures() };
+              QgsFeature selectedFeature;
+              while ( it.nextFeature( selectedFeature ) )
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmpackage.cpp:212]: Same here -- this shouldn't be run in the prepareAlgorithm stage.

**Comment by nyalldawson** [comment]
@elpaso 

FYI -- I'm hoping that for before GDAL 3.6 final I'll have the opportunity to extend https://github.com/OSGeo/gdal/pull/6104 with create support for relationships. We could then use that API to create relationship definitions in the GPKG which follow the related table extension specification (which QGIS will read and respect automatically following https://github.com/qgis/QGIS/pull/49504 )

**Comment by elpaso** [comment]
> @elpaso
> 
> FYI -- I'm hoping that for before GDAL 3.6 final I'll have the opportunity to extend [OSGeo/gdal#6104](https://github.com/OSGeo/gdal/pull/6104) with create support for relationships. We could then use that API to create relationship definitions in the GPKG which follow the related table extension specification (which QGIS will read and respect automatically following #49504 )

This would be great! Looking forward to it.

**Comment by elpaso** [comment]
unrelated test failure

**Pull request merged**